source/ada/lsp-ada_handlers.adb

```diff
@@ -431,14 +434,206 @@ package body LSP.Ada_Handlers is
+      Decl               : Basic_Decl;
+      Subp_Spec_Node     : Base_Subp_Spec;
+      Hover_Text         : LSP_String;
+
+      procedure Create_Hover_Text_For_Basic_Decl;
+      --  Create the hover text for for basic declarations
+
+      procedure Create_Hover_Text_For_Subp_Spec;
+      --  Create the hover text for for subprogram declarations
+
+      --------------------------------------
+      -- Create_Hover_Text_For_Basic_Decl --
+      --------------------------------------
+
+      procedure Create_Hover_Text_For_Basic_Decl is
+      begin
+         --  Return an empty hover text for package declarations
+         if Decl.Kind in Ada_Base_Package_Decl then
+            return;
+         end if;
+
+         declare
+            Text        : constant String := Langkit_Support.Text.To_UTF8
+              (Decl.Text);
+            Lines       : GNAT.Strings.String_List_Access := Split
+              (Text,
+               On               => ASCII.LF,
+               Omit_Empty_Lines => True);
+            Idx         : Integer;
+         begin
+            --  If it's a single-line declaration, replace all the series of
+            --  whitespaces by only one blankspace.
+            --  If it's a multi-line declaration, remove only the unneeded
+            --  indentation whitespaces.
+
+            if Lines'Length = 1 then
+               declare
+                  Res_Idx : Integer := Text'First;
+                  Result  : String (Text'First .. Text'Last);
+               begin
+                  Idx := Text'First;
+
+                  while Idx <= Text'Last loop
+                     Skip_Blanks (Text, Idx);
+
+                     while Idx <= Text'Last
+                       and then not Is_Whitespace (Text (Idx))
+                     loop
+                        Result (Res_Idx) := Text (Idx);
+                        Idx := Idx + 1;
+                        Res_Idx := Res_Idx + 1;
+                     end loop;
+
+                     if Res_Idx < Result'Last then
+                        Result (Res_Idx) := ' ';
+                        Res_Idx := Res_Idx + 1;
+                     end if;
+                  end loop;
+
+                  if Res_Idx > Text'First then
+                     Hover_Text := To_LSP_String
+                       (Result (Text'First .. Res_Idx - 1));
+                  end if;
+               end;
+            else
+               declare
+                  Blanks_Count_Per_Line : array
+                    (Lines'First + 1 .. Lines'Last) of Natural;
+                  Indent_Blanks_Count   : Natural := Natural'Last;
+               begin
+                  Hover_Text := To_LSP_String (Lines (Lines'First).all);
+
+                  --  Count the blankpaces per line and track how many
+                  --  blankspaces we should remove on each line by finding
+                  --  the common identation blankspaces.
+
+                  for J in Lines'First + 1 .. Lines'Last loop
```

Comment:
I may have missed something, but why do we not strip the leading blanks on the first line?

Reply:
Because the Text function returns the node's text starting from its first token, without the trivias before (e.g: for '   procedure Something is null' it will return 'procedure Something is null', without the leading blankspaces).